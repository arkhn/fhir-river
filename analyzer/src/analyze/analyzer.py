import re
import time
from collections.abc import Mapping

from analyzer.src.analyze.graphql import PyrogClient
from analyzer.src.config.logger import get_logger

from .analysis import Analysis
from .attribute import Attribute
from .cleaning_script import CleaningScript
from .concept_map import ConceptMap
from .condition import Condition
from .input_group import InputGroup
from .mapping import build_squash_rules
from .merging_script import MergingScript
from .sql_column import SqlColumn
from .sql_join import SqlJoin
from .sql_filter import SqlFilter

logger = get_logger()


class Analyzer:
    def __init__(self, pyrog_client: PyrogClient):
        self.pyrog = pyrog_client
        # Store analyses
        # TODO think about the design here. Use http caching instead of
        # storing here, for instance?
        self.analyses: Mapping = {}
        self.last_updated_at: Mapping = {}  # store last updated timestamp for each resource_id

        self._cur_analysis = Analysis()

    def get_analysis(self, resource_mapping_id, max_seconds_refresh=3600) -> Analysis:
        if resource_mapping_id not in self.analyses:
            self.fetch_analysis(resource_mapping_id)
        else:
            self.check_refresh_analysis(resource_mapping_id, max_seconds_refresh)
        return self.analyses[resource_mapping_id]

    def check_refresh_analysis(self, resource_mapping_id, max_seconds_refresh):
        """
        This method refreshes the analyser if the last update was later than `max_seconds_refresh`
        for each resource
        """
        if time.time() - self.last_updated_at.get(resource_mapping_id) > max_seconds_refresh:
            logger.info(
                f"Analysis too old for resource {resource_mapping_id}.",
                extra={"resource_id": resource_mapping_id},
            )
            self.fetch_analysis(resource_mapping_id)
        else:
            logger.info(
                "Analysis was updated recently. Using cached analysis.",
                extra={"resource_id": resource_mapping_id},
            )

    def fetch_analysis(self, resource_mapping_id):
        """
        Fetch mapping from API and store last updated timestamp
        :param resource_mapping_id:
        :return:
        """
        logger.info("Fetching mapping from api.", extra={"resource_id": resource_mapping_id})
        resource_mapping = self.pyrog.get_resource_from_id(resource_id=resource_mapping_id)
        self.analyze(resource_mapping)
        self.last_updated_at[resource_mapping_id] = time.time()

    # TODO add an update_analysis(self, resource_mapping_id)?

    def analyze(self, resource_mapping):
        self._cur_analysis = Analysis()

        # Analyze the mapping
        self.analyze_mapping(resource_mapping)

        if not self._cur_analysis.columns:
            self._cur_analysis.is_static = True
        else:
            # Add primary key to columns to fetch if needed
            self._cur_analysis.add_column(self._cur_analysis.primary_key_column)

            # Build squash rules
            self._cur_analysis.squash_rules = build_squash_rules(
                self._cur_analysis.columns,
                self._cur_analysis.joins,
                self._cur_analysis.primary_key_column.table_name(),
            )

        # Store analysis
        self.analyses[resource_mapping["id"]] = self._cur_analysis

        return self._cur_analysis

    def analyze_mapping(self, resource_mapping):
        self._cur_analysis.primary_key_column = self.get_primary_key(resource_mapping)
        self._cur_analysis.source_id = resource_mapping["source"]["id"]
        self._cur_analysis.source_credentials = resource_mapping["source"]["credential"]
        self._cur_analysis.resource_id = resource_mapping["id"]
        self._cur_analysis.definition_id = resource_mapping["definitionId"]
        self._cur_analysis.definition = resource_mapping["definition"]
        for filter_ in resource_mapping["filters"]:
            self.analyze_filter(filter_)
        for attribute_mapping in resource_mapping["attributes"]:
            self.analyze_attribute(attribute_mapping)

        return self._cur_analysis

    def analyze_filter(self, filter_):
        col = SqlColumn(
            filter_["sqlColumn"]["table"],
            filter_["sqlColumn"]["column"],
            self._cur_analysis.source_credentials["owner"],
        )
        sql_filter = SqlFilter(col, filter_["relation"], filter_["value"])
        self._cur_analysis.add_filter(sql_filter)

    def analyze_attribute(self, attribute_mapping):
        logger.info(
            f"Analyze attribute {attribute_mapping['path']} {attribute_mapping['definitionId']}",
            extra={"resource_id": self._cur_analysis.resource_id},
        )
        attribute = Attribute(
            path=attribute_mapping["path"], definition_id=attribute_mapping["definitionId"]
        )
        if not attribute_mapping["inputGroups"]:
            # If there are no input groups for this attribute, it means that it is an intermediary
            # attribute (ie not a leaf). It is here to give us some context information.
            # For instance, we can use it if its children attributes represent a Reference.
            if attribute_mapping["definitionId"] == "Reference":
                logger.info(
                    f"Analyze attribute reference",
                    extra={"resource_id": self._cur_analysis.resource_id},
                )
                # Remove trailing index
                path = re.sub(r"\[\d+\]$", "", attribute.path)
                self._cur_analysis.reference_paths.add(path)

            return

        for mapping_group in attribute_mapping["inputGroups"]:
            self.analyze_input_group(mapping_group, attribute)

        self._cur_analysis.attributes.append(attribute)

        return attribute

    def analyze_input_group(self, mapping_group, parent_attribute):
        input_group = InputGroup(id_=mapping_group["id"], attribute=parent_attribute)
        parent_attribute.add_input_group(input_group)
        for input_ in mapping_group["inputs"]:
            if input_["sqlValue"]:

                sqlValue = input_["sqlValue"]
                cur_col = SqlColumn(
                    sqlValue["table"],
                    sqlValue["column"],
                    self._cur_analysis.primary_key_column.owner,
                )

                if input_["script"]:
                    cur_col.cleaning_script = CleaningScript(input_["script"])

                if input_["conceptMapId"]:
                    cur_col.concept_map = ConceptMap(input_["conceptMapId"])

                for join in sqlValue["joins"]:
                    tables = join["tables"]
                    left = SqlColumn(
                        tables[0]["table"],
                        tables[0]["column"],
                        self._cur_analysis.primary_key_column.owner,
                    )
                    right = SqlColumn(
                        tables[1]["table"],
                        tables[1]["column"],
                        self._cur_analysis.primary_key_column.owner,
                    )
                    self._cur_analysis.add_join(SqlJoin(left, right))

                self._cur_analysis.add_column(cur_col)
                input_group.add_column(cur_col)

            elif input_["staticValue"]:
                input_group.add_static_input(input_["staticValue"])

        for mapping_condition in mapping_group["conditions"]:
            condition_column = SqlColumn(
                mapping_condition["sqlValue"]["table"],
                mapping_condition["sqlValue"]["column"],
                self._cur_analysis.primary_key_column.owner,
            )
            self._cur_analysis.add_column(condition_column)

            condition = Condition(
                action=mapping_condition["action"],
                sql_column=condition_column,
                value=mapping_condition["value"],
            )
            input_group.add_condition(condition)

        if mapping_group["mergingScript"]:
            input_group.merging_script = MergingScript(mapping_group["mergingScript"])

        return input_group

    def get_primary_key(self, resource_mapping):
        """ Get the primary key table and column of the provided resource.
        """
        if not resource_mapping["primaryKeyTable"] or not resource_mapping["primaryKeyColumn"]:
            raise ValueError(
                "You need to provide a primary key table and column in the mapping for "
                f"resource {resource_mapping['definitionId']}."
            )

        return SqlColumn(
            resource_mapping["primaryKeyTable"],
            resource_mapping["primaryKeyColumn"],
            resource_mapping["source"]["credential"]["owner"],
        )
