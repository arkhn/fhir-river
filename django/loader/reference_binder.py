import logging
import re
import json
from collections import defaultdict
from typing import DefaultDict
from dotty_dict import dotty

from arkhn_monitoring import Timer

from loader.load.utils import get_resource_id
from loader.cache import redis


logger = logging.getLogger(__name__)

# dotty-dict does not handle brackets indices,
# it uses dots instead (a.0.b instead of a[0].b)
def dotty_paths(paths):
    for path in paths:
        yield re.sub(r"\[(\d+)]", r".\1", path)


class ReferenceBinder:
    def __init__(self, fhirstore):
        self.fhirstore = fhirstore

        # In Redis, we have sets identified with keys defined as a stringified
        # json array:
        # "fhir_type_target:[value, system]"
        #
        # A Redis set is a list. Here a list of stringified json arrays
        # [
        #     "[ [fhir_type_source, path, isArray], fhir_id1 ]",
        #     "[ [fhir_type_source, path, isArray], fhir_id2 ]",
        #     ...
        # ]
        # eg: A Redis set "Practitioner:[1234, system]" contains
        # [
        #     "[ [Patient, generalPractitioner, True], fhir-pract-id1 ]",
        #     ...
        # ]
        self.cache = redis.conn()

    @Timer("time_resolve_references", "time spent resolving references")
    def resolve_references(self, unresolved_fhir_object, reference_paths):
        fhir_object = dotty(unresolved_fhir_object)
        resource_id = get_resource_id(unresolved_fhir_object)

        # iterate over the instance's references and try to resolve them
        for reference_path in dotty_paths(reference_paths):
            logger.debug(
                {
                    "message": f"Trying to resolve reference for resource {fhir_object['id']} "
                    f"at {reference_path}",
                    "resource_id": resource_id,
                },
            )
            try:
                bound_ref = self.bind_existing_reference(fhir_object, reference_path)
                fhir_object[reference_path] = bound_ref
            except KeyError as e:
                logger.warning(
                    f"{reference_path} does not exist in resource {fhir_object['id']}: {e}"
                )
        if "identifier" in fhir_object:
            self.resolve_pending_references(fhir_object)

        return fhir_object.to_dict()

    @Timer("time_bind_existing_reference", "time spent resolving the documents's references")
    def bind_existing_reference(self, fhir_object, reference_path):
        reference_attribute = fhir_object[reference_path]
        resource_id = get_resource_id(fhir_object)

        def bind(ref, is_array=False):
            # extract the type and itentifier of the reference
            reference_type = ref["type"]
            identifier = ref["identifier"]
            # search the referenced resource in the database
            try:
                _identifier = self.partial_identifier(identifier)
            except (ValueError, KeyError) as e:
                logger.warning(
                    f"incomplete identifier on reference of type "
                    f"{reference_type} at path {ref} of resource {fhir_object['id']}: {e}"
                )
                return ref
            referenced_resource = self.fhirstore.db[reference_type].find_one(_identifier, ["id"])
            if referenced_resource:
                # if found, add the ID as the "literal reference"
                # (https://www.hl7.org/fhir/references-definitions.html#Reference.reference)
                logger.debug(
                    {
                        "message": f"reference to {reference_type} {identifier} resolved",
                        "resource_id": resource_id,
                    },
                )
                ref["reference"] = f"{reference_type}/{referenced_resource['id']}"
            else:
                logger.debug(
                    {
                        "message": f"caching reference to {reference_type} {identifier} at {reference_path}",
                        "resource_id": resource_id,
                    },
                )
                target_ref = self.identifier_to_key(reference_type, identifier)
                source_ref = (fhir_object["resourceType"], reference_path, is_array)
                self.cache.sadd(target_ref, json.dumps((source_ref, fhir_object["id"])))
            return ref

        # If we have a list of references, we want to bind all of them.
        # Thus, we loop on all the items in reference_attribute.
        if isinstance(reference_attribute, list):
            return [bind(ref, is_array=True) for ref in reference_attribute]
        else:
            return bind(reference_attribute)

    @Timer("time_resolve_pending_references", "time spent resolving pending references")
    def resolve_pending_references(self, fhir_object):
        for identifier in fhir_object["identifier"]:
            try:
                target_ref = self.identifier_to_key(fhir_object["resourceType"], identifier)
            except (KeyError, ValueError) as e:
                logger.warning(
                    f"incomplete identifier on resource {fhir_object['id']}: {e}"
                )
                continue
            pending_refs = self.load_cached_references(target_ref)
            for (source_type, reference_path, is_array), refs in pending_refs.items():
                logger.debug(
                    {
                        "message": f"Updating {source_type} resources {', '.join(refs)} "
                        f"on reference {reference_path}",
                        "resource_id": get_resource_id(fhir_object),
                    },
                )
                self.fhirstore.db[source_type].update_many(
                    self.unresolved_resources_filter(reference_path, identifier, refs, is_array),
                    self.reference_update(reference_path, fhir_object, is_array)
                )
            if pending_refs:
                self.cache.delete(target_ref)

    @Timer("time_load_cached_references", "time spent loading references from redis")
    def load_cached_references(self, target_ref: str) -> DefaultDict[tuple, list]:
        """Requests cached references from Redis

        :param target_ref: "fhir_type_target:[value, system]"
        :type target_ref: str
        The SMEMBERS command gets the set target_ref and returns a list
        [
            "[ [fhir_type_source, path, isArray], fhir_id1 ]",
            "[ [fhir_type_source, path, isArray], fhir_id2 ]",
            ...
        ]
        This list is formatted as a defaultdict which is then returned
        :return pending_refs: {
            (fhir_type_source, path, isArray): [fhir_id1, fhir_id2],
            ...
        }
        :rtype: DefaultDict[tuple, list]
        """

        references_set = self.cache.smembers(target_ref)
        pending_refs = defaultdict(list)
        for element in references_set:
            (source_ref, ref) = json.loads(element)
            pending_refs[tuple(source_ref)].append(ref)
        return pending_refs

    def unresolved_resources_filter(self, reference_path, identifier, refs, is_array):
        query = {"id": {"$in": refs}}
        if is_array:
            query[reference_path] = {
                "$elemMatch": self.partial_identifier(identifier)
            }
        return query

    @staticmethod
    def reference_update(reference_path, fhir_object, is_array):
        if is_array:
            target_path = f"{reference_path}.$.reference"
        else:
            target_path = f"{reference_path}.reference"
        return {
            "$set": {
                target_path: f"{fhir_object['resourceType']}/{fhir_object['id']}"
            }
        }

    @staticmethod
    def identifier_to_key(resource_type, identifier):
        value = identifier["value"]
        system = identifier["system"]
        if not value or not system:
            raise ValueError
        # Default separators include whitespaces
        key = json.dumps((value, system), separators=(',', ':'))
        return f"{resource_type}:{key}"

    @staticmethod
    def partial_identifier(identifier):
        _id = {
            "identifier.value": identifier["value"],
            "identifier.system": identifier["system"]
        }
        if not identifier["value"] or not identifier["system"]:
            raise ValueError
        return _id
