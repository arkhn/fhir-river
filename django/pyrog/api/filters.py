from django_filters import rest_framework as filters
from pyrog import models


class CredentialFilterSet(filters.FilterSet):
    class Meta:
        model = models.Credential
        fields = ["source"]


class ColumnFilterSet(filters.FilterSet):
    class Meta:
        model = models.Column
        fields = ["join"]


class JoinFilterSet(filters.FilterSet):
    class Meta:
        model = models.Join
        fields = ["column"]


class FilterFilterSet(filters.FilterSet):
    class Meta:
        model = models.Filter
        fields = ["resource"]


class OwnerFilterSet(filters.FilterSet):
    class Meta:
        model = models.Owner
        fields = ["credential"]


class ResourceFilterSet(filters.FilterSet):
    class Meta:
        model = models.Resource
        fields = ["source"]


class AttributeFilterSet(filters.FilterSet):
    source = filters.ModelChoiceFilter("resource__source", queryset=models.Source.objects.all())

    class Meta:
        model = models.Attribute
        fields = ["resource", "source", "path"]


class InputGroupFilterSet(filters.FilterSet):
    class Meta:
        model = models.InputGroup
        fields = ["attribute"]


class StaticInputFilterSet(filters.FilterSet):
    class Meta:
        model = models.StaticInput
        fields = ["input_group"]


class SQLInputFilterSet(filters.FilterSet):
    class Meta:
        model = models.SQLInput
        fields = ["input_group"]


class ConditionFilterSet(filters.FilterSet):
    class Meta:
        model = models.Condition
        fields = ["input_group"]
