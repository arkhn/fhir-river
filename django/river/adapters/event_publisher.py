import dataclasses
import json
import logging

import confluent_kafka

from django.conf import settings

from river.domain.events import Event

logger = logging.getLogger(__name__)


class EventPublisher:
    def publish(self, topic: str, event: Event):
        raise NotImplementedError


class FakeEventPublisher(EventPublisher):
    def __init__(self):
        self._events: dict[str, set] = {}

    def publish(self, topic: str, event: Event):
        if topic in self._events:
            self._events[topic].add(event)
        else:
            self._events[topic] = {event}


class KafkaEventPublisher(EventPublisher):
    def __init__(
        self,
    ):
        self._kafka_producer = confluent_kafka.Producer(
            {
                "bootstrap.servers": settings.KAFKA_BOOTSTRAP_SERVERS,
                "linger.ms": 0.5,
                "session.timeout.ms": 6000,
            }
        )

    def publish(self, topic: str, event: Event):
        self._kafka_producer.produce(
            topic=topic,
            value=json.dumps(dataclasses.asdict(event)),
            callback=lambda err, msg: logger.debug("Message delivered")
            if err is None
            else logger.error(f"Message {event} delivery failed with error {err} for topic {msg.topic()}"),
        )
        self._kafka_producer.poll(1)
