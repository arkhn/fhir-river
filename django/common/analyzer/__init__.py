from .analyzer import Analyzer  # noqa
from .concept_map import ConceptMap  # noqa
