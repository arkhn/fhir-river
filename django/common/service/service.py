import logging
import re
from typing import Callable, Dict, Type

from confluent_kafka.error import ConsumeError

from river.adapters.event_subscriber import EventSubscriber
from river.domain import events

logger = logging.getLogger(__name__)
event_logger = logging.getLogger("river.event")


class Service:
    """Stateful service"""

    def __init__(self, subscriber: EventSubscriber, handlers: Dict[str, Callable[[Type[events.Event]], None]]) -> None:
        self.subscriber = subscriber
        self.handlers = handlers

    def run(self):
        topics = list(self.handlers.keys())
        with self.subscriber.subscribe(topics):
            while True:
                try:
                    topic, raw = self.subscriber.poll()
                except ConsumeError as err:
                    logger.error(err)
                    raise err

                try:
                    # topic can be regexp
                    topic = next(t for t in self.handlers.keys() if re.match(t, topic))
                except StopIteration:
                    logger.error(f"Unhandled topic: {topic}")

                try:
                    self.handlers[topic](raw)
                except Exception:
                    event_logger.exception("Failed to process event")
                    continue
