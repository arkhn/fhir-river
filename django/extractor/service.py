import logging

from django.conf import settings

import redis
from common.analyzer import Analyzer
from common.database_connection.db_connection import DBConnection
from common.kafka.consumer import Consumer
from common.kafka.producer import Producer
from common.service.event import Event
from common.service.handler import Handler
from common.service.service import Service
from confluent_kafka import KafkaError, KafkaException
from extractor.conf import conf
from extractor.extract import Extractor

logger = logging.getLogger(__name__)


def broadcast_events(
    dataframe,
    analysis,
    producer: Producer,
    counter_client: redis.Redis,
    received_event_data,
):
    resource_type = analysis.definition_id
    resource_id = analysis.resource_id
    batch_id = received_event_data["batch_id"]
    batch_type = received_event_data["batch_type"]
    count = 0
    # FIXME: Can we do better? It's to make sure this counter is ok
    # in the recurring case
    counter_client.hset(f"{batch_type}:{batch_id}:counter", f"resource:{resource_id}:loaded", 0)

    list_records_from_db = Extractor.split_dataframe(dataframe, analysis)
    for record in list_records_from_db:
        try:
            logger.debug(
                {"message": "One record from extract", "resource_id": resource_id},
            )
            event = dict()
            event["batch_id"] = batch_id
            event["batch_type"] = batch_type
            event["resource_type"] = resource_type
            event["resource_id"] = resource_id
            event["record"] = record
            producer.produce_event(
                topic=f"{conf.PRODUCED_TOPIC_PREFIX}.{batch_type}.{batch_id}",
                event=event,
            )
            count += 1
        except (KafkaException, ValueError) as err:
            if isinstance(err, KafkaException) and err.args[0].code() == KafkaError.UNKNOWN_TOPIC_OR_PART:
                logger.warning(
                    {
                        "message": "The current batch has been cancelled",
                        "resource_id": resource_id,
                        "batch_id": batch_id,
                    }
                )
            else:
                logger.exception(err)
            # return early to avoid setting the counter in redis
            return

    # Initialize a batch counter in Redis. For each resource_id, it sets
    # the number of produced records
    counter_client.hset(f"{batch_type}:{batch_id}:counter", f"resource:{resource_id}:extracted", count)
    logger.info(
        {
            "message": f"Batch {batch_id} size is {count} for resource type {analysis.definition_id}",
            "resource_id": resource_id,
        },
    )


class ExtractHandler(Handler):
    def __init__(
        self,
        producer: Producer,
        counter_redis: redis.Redis,
        analyzer: Analyzer,
    ) -> None:
        self.producer = producer
        self.counter_redis = counter_redis
        self.analyzer = analyzer

    def __call__(self, event: Event):
        batch_id = event.data["batch_id"]
        resource_id = event.data["resource_id"]
        primary_key_values = event.data.get("primary_key_values", None)

        analysis = self.analyzer.load_cached_analysis(batch_id, resource_id)
        db_connection = DBConnection(analysis.source_credentials)
        with db_connection.session_scope() as session:
            extractor = Extractor(session, db_connection.metadata)
            query = extractor.extract(analysis, primary_key_values)

            broadcast_events(query, analysis, self.producer, self.counter_redis, event.data)


class ExtractorService(Service):
    @classmethod
    def make_app(cls):
        config = {"max.poll.interval.ms": conf.MAX_POLL_INTERVAL_MS}

        consumer = Consumer(
            broker=settings.KAFKA_BOOTSTRAP_SERVERS,
            topics=conf.CONSUMED_TOPICS,
            group_id=conf.CONSUMER_GROUP_ID,
            config=config,
        )
        mapping_redis = redis.Redis(
            host=settings.REDIS_MAPPINGS_HOST,
            port=settings.REDIS_MAPPINGS_PORT,
            db=settings.REDIS_MAPPINGS_DB,
        )
        counter_redis = redis.Redis(
            host=settings.REDIS_COUNTER_HOST,
            port=settings.REDIS_COUNTER_PORT,
            db=settings.REDIS_COUNTER_DB,
        )
        analyzer = Analyzer(redis_client=mapping_redis)
        handler = ExtractHandler(
            producer=Producer(broker=settings.KAFKA_BOOTSTRAP_SERVERS),
            counter_redis=counter_redis,
            analyzer=analyzer,
        )
        return Service(consumer, handler)
