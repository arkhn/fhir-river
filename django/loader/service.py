import logging

from django.conf import settings

import redis

from common.service.application import Application
from common.service.event import Event
from common.service.handler import Handler
from common.kafka.consumer import Consumer
from common.kafka.producer import Producer
from common.analyzer import Analyzer

from loader.conf import conf
from loader.load import Loader
from loader.reference_binder import ReferenceBinder
from loader.load.fhirstore import get_fhirstore

logger = logging.getLogger(__name__)

from pymongo.errors import DuplicateKeyError


def load(
    fhir_object: dict,
    batch_id: str,
    resource_id: str,
    loader: Loader,
    producer: Producer,
    binder: ReferenceBinder,
    counter_redis: redis.Redis,
    analyzer: Analyzer,
):
    analysis = analyzer.load_cached_analysis(batch_id, resource_id)

    # Resolve existing and pending references (if the fhir_instance
    # references OR is referenced by other documents)
    logger.debug(
        f"Resolving references {analysis.reference_paths}",
        extra={"resource_id": resource_id},
    )
    resolved_fhir_instance = binder.resolve_references(
        fhir_object, analysis.reference_paths
    )

    try:
        logger.debug("Writing document to mongo", extra={"resource_id": resource_id})
        loader.load(
            resolved_fhir_instance,
            resource_type=resolved_fhir_instance["resourceType"],
        )
        # Increment loaded resources counter in Redis
        counter_redis.hincrby(
            f"batch:{batch_id}:counter", f"resource:{resource_id}:loaded", 1
        )
        producer.produce_event(
            topic=f"{conf.PRODUCED_TOPIC_PREFIX}{batch_id}",
            event={"batch_id": batch_id},
        )
    except DuplicateKeyError as err:
        logger.error(err)


class LoadHandler(Handler):
    def __init__(
        self,
        producer: Producer,
        loader: Loader,
        binder: ReferenceBinder,
        counter_redis: redis.Redis,
        analyzer: Analyzer,
    ) -> None:
        self.producer = producer
        self.loader = loader
        self.binder = binder
        self.counter_redis = counter_redis
        self.analyzer = analyzer

    def __call__(self, event: Event):
        load(
            producer=self.producer,
            loader=self.loader,
            binder=self.binder,
            counter_redis=self.counter_redis,
            analyzer=self.analyzer,
            **event.data,
        )


class LoaderApplication(Application):
    @classmethod
    def make_app(cls):
        fhirstore_client = get_fhirstore()
        loader = Loader(fhirstore_client)
        binder = ReferenceBinder(fhirstore_client)

        consumer = Consumer(
            broker=settings.KAFKA_BOOTSTRAP_SERVERS,
            topics=conf.CONSUMED_TOPICS,
            group_id=conf.CONSUMER_GROUP_ID,
        )
        mapping_redis = redis.Redis(
            host=settings.REDIS_MAPPINGS_HOST,
            port=settings.REDIS_MAPPINGS_PORT,
            db=settings.REDIS_MAPPINGS_DB,
        )
        counter_redis = redis.Redis(
            host=settings.REDIS_COUNTER_HOST,
            port=settings.REDIS_COUNTER_PORT,
            db=settings.REDIS_COUNTER_DB,
        )
        analyzer = Analyzer(redis_client=mapping_redis)
        handler = LoadHandler(
            producer=Producer(broker=settings.KAFKA_BOOTSTRAP_SERVERS),
            loader=loader,
            binder=binder,
            counter_redis=counter_redis,
            analyzer=analyzer,
        )
        return Application(consumer, handler)
