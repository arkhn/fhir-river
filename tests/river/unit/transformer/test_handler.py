import datetime

import pytest

from river.adapters.event_publisher import InMemoryEventPublisher
from river.common.analyzer import Analyzer
from river.domain.events import ExtractedRecord
from river.transformer.reference_binder import ReferenceBinder
from river.transformer.service import extracted_record_handler
from river.transformer.transformer import Transformer
from syrupy.filters import paths

pytestmark = pytest.mark.django_db


def test_extracted_resource_handler(batch_factory, mimic_mapping, snapshot):
    # FIXME: use a dedicated fixture for the patient mapping
    # instead of the first resource of mimic mappings.
    resource_id = "ckt4kpz1800u63hvz7p1wf9xi"

    batch = batch_factory.create(id="test-batch-id")
    event = ExtractedRecord(
        batch_id=batch.id,
        resource_type="",
        resource_id=resource_id,
        record={
            "patients_subject_id_a9fb9667": ["didier@chloroquine.org"],
            "patients_gender_8035bbd1": ["F"],
            "patients_dob_f873eb32": [datetime.datetime(2057, 11, 15, 0, 0)],
            "patients_dod_9fd82e01": [datetime.datetime(2114, 2, 20, 0, 0)],
        },
    )
    publisher = InMemoryEventPublisher()
    analyzer = Analyzer()
    transformer = Transformer()
    binder = ReferenceBinder()

    extracted_record_handler(
        event=event,
        publisher=publisher,
        analyzer=analyzer,
        transformer=transformer,
        binder=binder,
    )

    assert f"transform.{batch.id}" in publisher._events
    assert publisher._events[f"transform.{batch.id}"][0].fhir_object == snapshot(exclude=paths("meta.lastUpdated"))
