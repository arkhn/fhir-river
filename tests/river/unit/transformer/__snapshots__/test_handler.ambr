# name: test_extracted_resource_handler
  <class 'dict'> {
    'birthDate': '2057-11-15',
    'communication': <class 'list'> [
      <class 'dict'> {
        'language': <class 'dict'> {
          'coding': <class 'list'> [
            <class 'dict'> {
              'code': 'RUSS',
            },
            <class 'dict'> {
              'code': 'ENGL',
            },
          ],
        },
      },
    ],
    'deceasedDateTime': '2114-02-20',
    'gender': 'female',
    'id': '50fe8f2e-bd34-561c-91b3-055813fba3b8',
    'identifier': <class 'list'> [
      <class 'dict'> {
        'value': '31789 (1)',
      },
    ],
    'maritalStatus': <class 'dict'> {
      'coding': <class 'list'> [
        <class 'dict'> {
          'code': 'SINGLE',
        },
      ],
    },
    'meta': <class 'dict'> {
      'tag': <class 'list'> [
        <class 'dict'> {
          'code': 'cktlnp0bb00020mmz5fks20ij',
          'system': 'http://terminology.arkhn.org/CodeSystem/project',
        },
        <class 'dict'> {
          'code': 'cktlnp0ji006e0mmzat7dwb98',
          'system': 'http://terminology.arkhn.org/CodeSystem/resource',
        },
      ],
    },
    'resourceType': 'Patient',
  }
---
